### PR TITLE
Cw/grokking algos

### PR DESCRIPTION
**Pull request by Cam-Wheeler**
Hello Mattia, here are my changes. 

src/TD.py
- Addition of Sarsa Lambda 
- Addition of Watkins Q Learning

examples/TD.py
-Addition of code that runs both algorithms. 

Algorithms run as expected, Watkins Q produces really good policies, Sarsa lambda not so much. 
Currently working on a model based algorithm, but it will take some more time!

### CODE REVIEW

File: algorl/src/TD.py [hunk 1]
```diff
@@ -87,7 +90,10 @@ def td_control(self, algo, plot_name:str):
                 reward = self.env.grid[next_state]
 
                 # Compute the pair state-actions
-                q_values_df = algo(q_values_df, state, action, reward, next_state, next_action)
+                if lambda_target == False:
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:93]: You can replace it with 
``` if lambda_target:```

File: algorl/src/TD.py [hunk 2]
```diff
@@ -530,4 +539,113 @@ def compute_state_values(self):
             self.env.grid[grid_state]=state_values[state_values.states == grid_state].state_value
                 
 
-        
\ No newline at end of file
+class Sarsa_Lambda(TemporalDifferenceFunctions):
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:542]: Class names want camel case, replace Sarsa_Lambda with
```SarsaLambda```

File: algorl/src/TD.py [hunk 3]
```diff
@@ -530,4 +539,113 @@ def compute_state_values(self):
...
+    Trace vector can be set as either accumulating or replacing. 
+    Policy is epsilon-greedy. 
+    Reference:
+    --------------------
+    - Grokking Deep Reinforcement Learning by Miguel Morales. Page 210.
+    """
+
+    def __init__(
+            self, env, alpha:float = 0.5, gamma:float = 0.9, lambda_:float = 0.5, 
+            epsilon:bool = 0.1, replacing_traces:bool = False, starting_state = (2,0),
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:554]: epsilon should be a float not bool
starting state a tuple of integers.
Check hints documentation for more

File: algorl/src/TD.py [hunk 4]
```diff
@@ -530,4 +539,113 @@ def compute_state_values(self):
...
+    Reference:
+    --------------------
+    - Grokking Deep Reinforcement Learning by Miguel Morales. Page 210.
+    """
+
+    def __init__(
+            self, env, alpha:float = 0.5, gamma:float = 0.9, lambda_:float = 0.5, 
+            epsilon:bool = 0.1, replacing_traces:bool = False, starting_state = (2,0), 
+            num_of_epochs:int = 1_000, num_episodes = 10_000, plot_name = 'Sarsa_Lambda', 
+            reward = -1):
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:556]: Same here, int, str and int, missing.
You can also try to add the expected result(s) using '->'
check the decumentation

File: algorl/src/TD.py [hunk 5]
```diff
@@ -530,4 +539,113 @@ def compute_state_values(self):
...
+
+        return q_values_df, eligibility_df
+
+    def compute_state_value(self, plot_name='SARSA-Lambda'):
+        self.logger.info('Compute SARSA Lambda')
+        self.td_control(algo=self.sarsa_lambda_formula, plot_name=plot_name, lambda_target=True)
+
+    
+
+class Watkins_Q(TemporalDifferenceFunctions): # AKA Q Learning Lambda.
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:597]: ``` WalkinsQ```

File: algorl/src/TD.py [hunk 6]
```diff
@@ -530,4 +539,113 @@ def compute_state_values(self):
...
+    Reference:
+    --------------------
+    - Grokking Deep Reinforcement Learning by Miguel Morales. Page 214.     
+    """
+
+    def __init__(
+            self, env, alpha:float = 0.5, gamma:float = 0.9, lambda_:float = 0.5, 
+            epsilon:bool = 0.1, replacing_traces:bool = False, starting_state = (2,0), 
+            num_of_epochs:int = 1_000, num_episodes = 10_000, plot_name = 'Watkins Q', 
+            reward = -1):
```
Comment (MattiaCinelli) [inline algorl/src/TD.py:610]: fix the hints

**Review by MattiaCinelli — CHANGES_REQUESTED** [review]
The image result for Walkins is perfect. 
Sarsa lambda not so much, please double check any possible error if not found, well, it is what it is.

For the rest all good just fix the hints and the Camel case

File: algorl/src/TD.py (hunk 1 above)
Comment (CamWheeler135) [reply algorl/src/TD.py:93]: Would you like to code to run like this: 

if lambda_target:
        q_values_df, eligibility_df = algo(q_values_df, eligibility_df, state, action, reward, next_state, next_action)
else:
        q_values_df = algo(q_values_df, state, action, reward, next_state, next_action)

**Comment by CamWheeler135** [comment]
I have just pushed the fixes for the camelcase and hints, I will check the code for SARSA Lambda and look for errors next.

**Review by MattiaCinelli — APPROVED** [review]

**Comment by CamWheeler135** [comment]
Hello!

I went back to look at the code again and it appears to be okay! I wanted to try and find out why SARSA was underperforming and I found that because it is "on-policy" it really struggles to estimate the optimal action-value function. For reference please see Grokking DRL pages 230, it really shows the difference between how fast Watkins Q can estimate vs SARSA(lambda), which does not converge at all on the 4x4 frozen lake with 10,000 episodes.

**Pull request merged**